### PR TITLE
Implement Inclavare Containers PAL interface in WAMR Linux-SGX

### PR DESCRIPTION
**Pull request by leyao-daily**
This patch is the implemtation of the PAL interface [spec_v2](https://github.com/alibaba/inclavare-containers/blob/master/rune/libenclave/internal/runtime/pal/spec_v2.md)
And it enables WAMR as a enclave runtime backend for Inclavare Containers.
I also draft a guide for how to use it.

In the future, I will add extra arguments support and divide the workflow into the standard PAL processions.

### CODE REVIEW

**Comment by xwang98** [comment]
@leyao-daily This is great work!

File: product-mini/platforms/linux-sgx/enclave-sample/App/App.cpp [hunk 1]
```diff
@@ -734,3 +739,136 @@ main(int argc, char *argv[])
...
+                printf("%s\n", error_buf);
+		printf("Failed to load buffer to wams module\n");
+                return -1;
+        }
+
+        /* Set wasi arguments */
+        if (!set_wasi_args(wasm_module, dir_list, dir_list_size,
+                               env_list, env_list_size, argv, argc)) {
+                printf("%s\n", "set wasi arguments failed.\n");
+		printf("Failed to set wasi arguments\n");
```
Comment (wenyongh) [inline product-mini/platforms/linux-sgx/enclave-sample/App/App.cpp:820]: Same as above

File: product-mini/platforms/linux-sgx/enclave-sample/App/App.cpp [hunk 2]
```diff
@@ -734,3 +739,136 @@ main(int argc, char *argv[])
...
+                return -1;
+        }
+
+        /* Instantiate module */
+        if (!(wasm_module_inst[i] = instantiate_module(wasm_module,
+                                                stack_size, heap_size,
+                                                error_buf,
+                                                sizeof(error_buf)))) {
+                printf("%s\n", error_buf);
+		printf("Failed to read file to buffer\n");
```
Comment (wenyongh) [inline product-mini/platforms/linux-sgx/enclave-sample/App/App.cpp:830]: indent
unload wasm_module, free file buffer, destroy runtime, and return?

File: product-mini/platforms/linux-sgx/enclave-sample/App/App.cpp [hunk 3]
```diff
@@ -734,3 +739,136 @@ main(int argc, char *argv[])
...
+                return -1;
+        }
+
+
+        app_instance_main(wasm_module_inst[i], argc, argv);
+
+        /* Deinstantiate module */
+        deinstantiate_module(wasm_module_inst[i]);
```
Comment (wenyongh) [inline product-mini/platforms/linux-sgx/enclave-sample/App/App.cpp:838]: Should also unload module, free file buffer after deinstantiate module?

File: product-mini/platforms/linux-sgx/enclave-sample/App/App.cpp [hunk 4]
```diff
@@ -734,3 +739,136 @@ main(int argc, char *argv[])
...
+    }
+    return 0;
```
Comment (wenyongh) [inline product-mini/platforms/linux-sgx/enclave-sample/App/App.cpp:840]: Should destroy runtime before return?

File: product-mini/platforms/linux-sgx/enclave-sample/App/App.cpp [hunk 5]
```diff
@@ -734,3 +739,136 @@ main(int argc, char *argv[])
...
+}
+
+
+int wamr_pal_destroy(void) {
+    destroy_runtime();
```
Comment (wenyongh) [inline product-mini/platforms/linux-sgx/enclave-sample/App/App.cpp:845]: if wamr_pal_create_process destroys runtime, here we don't need to destroy again.

File: product-mini/platforms/linux-sgx/enclave-sample/App/App.cpp [hunk 6]
```diff
@@ -734,3 +739,136 @@ main(int argc, char *argv[])
     return 0;
 }
 
+int wamr_pal_get_version(void) {
```
Comment (wenyongh) [inline product-mini/platforms/linux-sgx/enclave-sample/App/App.cpp:742]: Move '{' to next line

File: product-mini/platforms/linux-sgx/enclave-sample/App/App.cpp [hunk 7]
```diff
@@ -734,3 +739,136 @@ main(int argc, char *argv[])
...
+
+int wamr_pal_destroy(void) {
+    destroy_runtime();
+    //sgx_destroy_enclave(g_eid);
+    return 0;
+}
+
+int wamr_pal_exec(struct wamr_pal_exec_args *args) { 
+    //app_instance_main(wasm_module_inst[i], argc, argv);
+    return 0; }
```
Comment (wenyongh) [inline product-mini/platforms/linux-sgx/enclave-sample/App/App.cpp:852]: Move '}' to next line

File: product-mini/platforms/linux-sgx/enclave-sample/App/pal_api.h [hunk 8]
```diff
@@ -0,0 +1,145 @@
+#ifndef __WAMR_PAL_API_H__
```
[the remark this answers is not in the document]
Comment (leyao-daily) [reply product-mini/platforms/linux-sgx/enclave-sample/App/pal_api.h:1]: Add it.

File: product-mini/platforms/linux-sgx/enclave-sample/App/App.cpp (hunk 5 above)
Comment (leyao-daily) [reply product-mini/platforms/linux-sgx/enclave-sample/App/App.cpp:845]: We need destroy the runtime in wamr_pal_destroy, so in pal_create_process we only free the file buffer and unload the module.

File: product-mini/platforms/linux-sgx/enclave-sample/App/App.cpp [hunk 9]
```diff
@@ -734,3 +739,136 @@ main(int argc, char *argv[])
...
+    uint8_t *wasm_files_buf = NULL;
+    void *wasm_modules = NULL;
+    int len=0;
+    
+    char *temp = (char *)args->argv[0];
+    while(temp){
+        len++;
+        temp=(char *)args->argv[len];
+    }
+    char *wasm_files[len];
```
[the remark this answers is not in the document]
Comment (leyao-daily) [reply product-mini/platforms/linux-sgx/enclave-sample/App/App.cpp]: Okay. I changed it

File: product-mini/platforms/linux-sgx/enclave-sample/App/App.cpp [hunk 10]
```diff
@@ -734,3 +739,136 @@ main(int argc, char *argv[])
...
+    for (int i=0; i<len; ++i) {
+        uint8_t *wasm_file_buf = NULL;
+        uint32_t wasm_file_size;
+        void *wasm_module = NULL;
+        char error_buf[128] = { 0 };
+        /* Load WASM byte buffer from WASM bin file */
+        if (!(wasm_file_buf =
+            (uint8_t *)read_file_to_buffer(wasm_files[i], &wasm_file_size))) {
+                printf("Failed to read file to buffer\n");
+                return -1;
```
[the remark this answers is not in the document]
Comment (leyao-daily) [reply product-mini/platforms/linux-sgx/enclave-sample/App/App.cpp]: The destroy should be called by pal_destroy, so remove from here.

**Pull request merged**